sklearn/decomposition/dict_learning.py

```diff
@@ -472,98 +501,39 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
-        code = np.c_[code, np.zeros((len(code), n_components - r))]
-        dictionary = np.r_[dictionary,
-                           np.zeros((n_components - r, dictionary.shape[1]))]
-
-    # Fortran-order dict, as we are going to access its row vectors
-    dictionary = np.array(dictionary, order='F')
-
-    residuals = 0
-
-    errors = []
-    current_cost = np.nan
-
-    if verbose == 1:
-        print('[dict_learning]', end=' ')
-
-    # If max_iter is 0, number of iterations returned should be zero
-    ii = -1
-
-    for ii in range(max_iter):
-        dt = (time.time() - t0)
-        if verbose == 1:
-            sys.stdout.write(".")
-            sys.stdout.flush()
-        elif verbose:
-            print("Iteration % 3i "
-                  "(elapsed time: % 3is, % 4.1fmn, current cost % 7.3f)"
-                  % (ii, dt, dt / 60, current_cost))
-
-        # Update code
-        code = sparse_encode(X, dictionary, algorithm=method, alpha=alpha,
-                             init=code, n_jobs=n_jobs)
-        # Update dictionary
-        dictionary, residuals = _update_dict(dictionary.T, X.T, code.T,
-                                             verbose=verbose, return_r2=True,
-                                             random_state=random_state)
-        dictionary = dictionary.T
-
-        # Cost function
-        current_cost = 0.5 * residuals + alpha * np.sum(np.abs(code))
-        errors.append(current_cost)
-
-        if ii > 0:
-            dE = errors[-2] - errors[-1]
-            # assert(dE >= -tol * errors[-1])
-            if dE < tol * errors[-1]:
-                if verbose == 1:
-                    # A line return
-                    print("")
-                elif verbose:
-                    print("--- Convergence reached after %d iterations" % ii)
-                break
-        if ii % 5 == 0 and callback is not None:
-            callback(locals())
-
-    if return_n_iter:
-        return code, dictionary, errors, ii + 1
-    else:
-        return code, dictionary, errors
+    Parameters
+    ==========
+    X: ndarray, shape (n_samples, n_features)
+        The input data.
+    V: ndarray, shape (n_features, n_components)
+        The dictionary.
+    U: ndarray, shape (n_samples, n_components)
+        The codes.
+    """
+    residuals = V.dot(U)
+    residuals -= X.T
+    residuals **= 2
+    residuals = np.sum(residuals)
+    return residuals
 
 
 def dict_learning_online(X, n_components=2, alpha=1, n_iter=100,
                          return_code=True, dict_init=None, callback=None,
-                         batch_size=3, verbose=False, shuffle=True, n_jobs=1,
-                         method='lars', iter_offset=0, random_state=None,
-                         return_inner_stats=False, inner_stats=None,
-                         return_n_iter=False):
+                         batch_size=None, verbose=False, shuffle=True,
+                         n_jobs=1, method='lars', iter_offset=0, tol=0.,
```

Comment:
functions starting with "_" are not documented and not visible to the user normally, so it will be just a way for us to organize the implementation.

I think that's preferable to having parameters that only work some of the time.